pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1163 @@
+			let acc = ensure_signed(origin)?;
+			ensure!(!Self::is_candidate(&acc), Error::<T>::CandidateExists);
+			ensure!(!Self::is_delegator(&acc), Error::<T>::DelegatorExists);
+			ensure!(bond >= T::MinCollatorCandidateStk::get(), Error::<T>::ValBondBelowMin);
+			ensure!(bond <= T::MaxCollatorCandidateStk::get(), Error::<T>::ValBondAboveMax);
+
+			let mut candidates = <CandidatePool<T>>::get();
+			ensure!(
+				candidates.insert(Bond {
+					owner: acc.clone(),
+					amount: bond
+				}),
+				Error::<T>::CandidateExists
+			);
+
+			// Post-launch TODO: Replace with `check_collator_candidate_inclusion`.
+			ensure!(
+				(candidates.len() as u32) <= T::MaxCollatorCandidates::get(),
+				Error::<T>::TooManyCollatorCandidates
+			);
+			T::Currency::reserve(&acc, bond)?;
+
+			let candidate = Collator::new(acc.clone(), bond);
+			let TotalStake {
+				collators: total_collators,
+				delegators: total_delegators,
+			} = <Total<T>>::get();
+			let total_collators = total_collators.saturating_add(bond);
+			<Total<T>>::put(TotalStake {
+				collators: total_collators,
+				delegators: total_delegators,
+			});
+			<CollatorState<T>>::insert(&acc, candidate);
+			<CandidatePool<T>>::put(candidates);
+
+			Self::deposit_event(Event::JoinedCollatorCandidates(acc, bond, total_collators));
+			Ok(().into())
+		}
+
+		/// Request to leave the set of candidates. If successful, the account
+		/// is immediately removed from the candidate pool to prevent selection
+		/// as a collator, but unbonding is executed with a delay of
+		/// `BondDuration` rounds.
+		#[pallet::weight(0)]
+		pub fn leave_candidates(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let collator = ensure_signed(origin)?;
+			let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			ensure!(!state.is_leaving(), Error::<T>::AlreadyLeaving);
+			let mut exits = <ExitQueue<T>>::get();
+			let now = <Round<T>>::get().current;
+			let when = now.saturating_add(T::BondDuration::get());
+			ensure!(
+				exits.insert(Bond {
+					owner: collator.clone(),
+					amount: when
+				}),
+				Error::<T>::AlreadyLeaving
+			);
+			state.leave_candidates(when);
+			let mut candidates = <CandidatePool<T>>::get();
+			if candidates.remove_by(|bond| bond.owner.cmp(&collator)).is_some() {
+				<CandidatePool<T>>::put(candidates);
+			}
+			<ExitQueue<T>>::put(exits);
+			<CollatorState<T>>::insert(&collator, state);
+			Self::deposit_event(Event::CollatorScheduledExit(now, collator, when));
+			Ok(().into())
+		}
+
+		/// Temporarily leave the set of collator candidates without unbonding
+		#[pallet::weight(0)]
+		pub fn go_offline(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let collator = ensure_signed(origin)?;
+			let mut state = <CollatorState<T>>::get(&collator).ok_or(Error::<T>::CandidateDNE)?;
+			ensure!(state.is_active(), Error::<T>::AlreadyOffline);
```

Comment:
What if state is `is_leaving`? Should that trigger an AlreadyOffline error? Might be better something like NotActive, or something similar to what is done in `go_online()`

Reply:
Good catch! Fixed in 58dcb24ffb425040c499fb2c2e292287a2d9105c